### PR TITLE
Commit allows turning gtags support off

### PR DESCRIPTION
**Pull request by jarvisschultz**
GNU global/ggtags-mode support was introduced in #267. This works well, but doesn't give the user the ability to control whether to use ggtags-mode or not. I imagine use cases where users would want to have the ability to control which backend to use for `projectile-regenerate-tags` and `projectile-find-tag` on a project-by-project basis.

### CODE REVIEW

File: projectile.el [hunk 1]
```diff
@@ -2827,7 +2835,7 @@ entirely."
   :group 'projectile
   :type 'sexp
   :risky t
-  :package-version '(projectile "0.12.0"))
+  :package-version '(projectile "0.13.0"))
```
Comment (bbatsov) [inline projectile.el]: This change looks unrelated.

File: projectile.el [hunk 2]
```diff
@@ -363,7 +369,7 @@ If a buffer is in the list projectile will ignore
 it for functions working with buffers."
   :group 'projectile
   :type '(repeat string)
-  :package-version '(projectile . "0.12.0"))
+  :package-version '(projectile . "0.13.0"))
```
Comment (bbatsov) [inline projectile.el]: This change looks unrelated.

File: projectile.el [hunk 3]
```diff
@@ -211,6 +211,12 @@ Otherwise consider the current directory the project root."
   :group 'projectile
   :type 'string)
 
+(defcustom projectile-use-gtags nil
```
Comment (bbatsov) [inline projectile.el]: I think we should simply give the users the option to explicitly select what they want to use - gtags, etags-select or find-tag. Boolean values are not a good idea when you have multiple options.

File: projectile.el [hunk 4]
```diff
@@ -211,6 +211,12 @@ Otherwise consider the current directory the project root."
...
+  "If true, Projectile will use GNU global via `ggtags-mode'
```
Comment (bbatsov) [inline projectile.el]: The first line of any docstring should be a complete sentence.

File: projectile.el [hunk 5]
```diff
@@ -211,6 +211,12 @@ Otherwise consider the current directory the project root."
...
+instead of ctags/etags for generating and finding tags."
+  :group 'projectile
+  :type 'boolean)
```
Comment (bbatsov) [inline projectile.el]: This is missing a package-version.

**Comment by bbatsov** [comment]
Apart from my inline remarks - please, take a look at our commit message guidelines and add a changelog entry.

**Comment by bbatsov** [comment]
ping

**Comment by jarvisschultz** [comment]
Sorry... haven't had time to update this PR. All of your suggestions make sense and should be easy to incorporate. I'll try to get to it soon.

**Comment by jarvisschultz** [comment]
I tried to address all comments. I'm no elisp expert... if something looks really stupid, I wouldn't be surprised. Don't hesitate to call me out.

I didn't add a `package-version` tag to the `defcustom` call because I wasn't sure what version should be assigned (0.14?). 

I tried to preserve the old behavior with the "auto" setting.

**Comment by bbatsov** [comment]
> I didn't add a package-version tag to the defcustom call because I wasn't sure what version should be assigned (0.14?).

Yeah, it will be 0.14.

File: projectile.el [hunk 6]
```diff
@@ -211,6 +211,27 @@ Otherwise consider the current directory the project root."
...
+
+If set to 'auto', `projectile-find-tag' will automatically choose
+which backend to use. Preference order is ggtags -> etags-select
+-> find-tag. Variable can also be set to specify which backend to
+use. If selected backend is unavailable, fall back to `find-tag'.
+
+If this variable is set to 'auto' and ggtags is available, or if
+set to 'ggtags', then ggtags will be used for
+`projectile-regenerate-tags'. For all other settings
+`projectile-tags-command' will be used.
```
Comment (bbatsov) [inline projectile.el]: Move the closing " to the previous line.

File: projectile.el [hunk 7]
```diff
@@ -211,6 +211,27 @@ Otherwise consider the current directory the project root."
...
+"
+  :group 'projectile
+  :type '(radio
+          (const :tag "auto" auto)
+          (const :tag "ggtags" ggtags)
+          (const :tag "etags" etags-select)
+          (const :tag "standard" find-tag)
+           ))
```
Comment (bbatsov) [inline projectile.el]: Move all closing parens to the previous line.

File: projectile.el [hunk 8]
```diff
@@ -1949,7 +1970,9 @@ regular expression."
 (defun projectile-regenerate-tags ()
   "Regenerate the project's [e|g]tags."
   (interactive)
-  (if (boundp 'ggtags-mode)
+  (if (and (boundp 'ggtags-mode)
+        (or (eq projectile-tags-backend 'auto)
```
Comment (bbatsov) [inline projectile.el]: The indentation of this expression is incorrect. You can simplify it a bit by using `memq`.

File: projectile.el [hunk 9]
```diff
@@ -1984,13 +2007,30 @@ regular expression."
...
   ;; Auto-discover the user's preference for tags
-  (let ((find-tag-fn (cond
-                      ((fboundp 'ggtags-find-tag-dwim)
-                       'ggtags-find-tag-dwim)
-                      ((fboundp 'etags-select-find-tag)
-                       'etags-select-find-tag)
-                      (t
-                       'find-tag))))
+  (let ((find-tag-fn
+          (cond
```
Comment (bbatsov) [inline projectile.el]: I'd extract the big `cond` into a separate function.

File: projectile.el [hunk 10]
```diff
@@ -1984,13 +2007,30 @@ regular expression."
...
+            ((eq projectile-tags-backend 'auto)
+              (cond
+                ((fboundp 'ggtags-find-tag-dwim)
+                  'ggtags-find-tag-dwim)
+                ((fboundp 'etags-select-find-tag)
+                  'etags-select-find-tag)
+                (t
+                  'find-tag)))
+            ((eq projectile-tags-backend 'ggtags)
+              (cond
```
Comment (bbatsov) [inline projectile.el]: You can simply use `if` here.

File: projectile.el [hunk 11]
```diff
@@ -1984,13 +2007,30 @@ regular expression."
...
+                ((fboundp 'ggtags-find-tag-dwim)
+                  'ggtags-find-tag-dwim)
+                (t
+                  'find-tag)))
+            ((eq projectile-tags-backend 'etags)
+              (cond
```
Comment (bbatsov) [inline projectile.el]: Same here.

**Comment by bbatsov** [comment]
I've added a few more minor remarks. After you address them squash all the commits into one.

**Comment by bbatsov** [comment]
@jarvisschultz ping :-)

**Comment by jarvisschultz** [comment]
Ugh... I'm the worst... sorry I've been on the road for work and haven't updated this. Should only take a few minutes, but I don't have my development machine, and I'm out of contact till late next week.

I haven't forgotten :)

**Comment by jarvisschultz** [comment]
Finally got around to this. So sorry it took me so long. I believe I addressed all previous comments, and all commits have been squashed to a single commit that is up-to-date with current master branch. Thanks for all the help!
